{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neostats import ort, stats, uni\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0 • CONFIG •"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "########## 12. Statistics\n",
    "\n",
    "rf = ['Mammalia', 'Sauropsida', 'Actinopteri']             # select a reference classes list, None if you don't want to choose one           \n",
    "rs = ['Homo_sapiens', 'Gallus_gallus', 'Danio_rerio']      # select a reference species list\n",
    "du = ['tandem', 'convergent', 'divergent']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Orthologues accessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Tandem'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ort.whichdup('ENSP00000256999')                         # Return duplication kind for a given ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>0A</th>\n",
       "      <th>0B</th>\n",
       "      <th>1A</th>\n",
       "      <th>1B</th>\n",
       "      <th>2A</th>\n",
       "      <th>2B</th>\n",
       "      <th>3A</th>\n",
       "      <th>3B</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>...</th>\n",
       "      <th>982A</th>\n",
       "      <th>982B</th>\n",
       "      <th>983A</th>\n",
       "      <th>983B</th>\n",
       "      <th>984A</th>\n",
       "      <th>984B</th>\n",
       "      <th>985A</th>\n",
       "      <th>985B</th>\n",
       "      <th>986A</th>\n",
       "      <th>986B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Classes</th>\n",
       "      <th>Orders</th>\n",
       "      <th>Species</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Actinopteri</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Anabantiformes</th>\n",
       "      <th>Anabas_testudineus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSATEP00000010433</td>\n",
       "      <td>ENSATEP00000011421</td>\n",
       "      <td>ENSATEP00000026918</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSATEP00000025936</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSATEP00000031779</td>\n",
       "      <td>ENSATEP00000031806</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSATEP00000005563</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSATEP00000014120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Betta_splendens</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSBSLP00000049808</td>\n",
       "      <td>ENSBSLP00000036647</td>\n",
       "      <td>ENSBSLP00000013895</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSBSLP00000019183</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSBSLP00000000958</td>\n",
       "      <td>ENSBSLP00000000940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSBSLP00000037416</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSBSLP00000026647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Beloniformes</th>\n",
       "      <th>Oryzias_latipes</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSORLP00000022912</td>\n",
       "      <td>ENSORLP00000037782</td>\n",
       "      <td>ENSORLP00000007185</td>\n",
       "      <td>ENSORLP00000045928</td>\n",
       "      <td>ENSORLP00000034755</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSORLP00000002323</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSORLP00000001643</td>\n",
       "      <td>ENSORLP00000038989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSORLP00000013968</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSORLP00000025592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Blenniiformes</th>\n",
       "      <th>Gouania_willdenowi</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGWIP00000010052</td>\n",
       "      <td>ENSGWIP00000042079</td>\n",
       "      <td>ENSGWIP00000052479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGWIP00000054609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSGWIP00000010024</td>\n",
       "      <td>ENSGWIP00000010029</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGWIP00000052240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGWIP00000014315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Salarias_fasciatus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSSFAP00005052284</td>\n",
       "      <td>ENSSFAP00005039658</td>\n",
       "      <td>ENSSFAP00005007557</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSSFAP00005037587</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSSFAP00005047708</td>\n",
       "      <td>ENSSFAP00005047718</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSSFAP00005047481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Sauropsida</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">Testudines</th>\n",
       "      <th>Gopherus_agassizii</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGAGP00000032669</td>\n",
       "      <td>ENSGAGP00000002463</td>\n",
       "      <td>ENSGAGP00000015760</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGAGP00000034764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGAGP00000003994</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGAGP00000004118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSGAGP00000027308</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pelodiscus_sinensis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPSIP00000004165</td>\n",
       "      <td>ENSPSIP00000017723</td>\n",
       "      <td>ENSPSIP00000008342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPSIP00000010788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPSIP00000017605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPSIP00000004610</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pelusios_castaneus</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPCEP00000004116</td>\n",
       "      <td>ENSPCEP00000016745</td>\n",
       "      <td>ENSPCEP00000021009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPCEP00000015691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPCEP00000009535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPCEP00000015385</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSPCEP00000007761</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Terrapene_carolina_triunguis</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTMTP00000031552</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTMTP00000023774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTMTP00000030499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTMTP00000026744</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSTMTP00000000781</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tinamiformes</th>\n",
       "      <th>Nothoprocta_perdicaria</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSNPEP00000008724</td>\n",
       "      <td>ENSNPEP00000013660</td>\n",
       "      <td>ENSNPEP00000006243</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSNPEP00000021207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168 rows × 1974 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          0A  \\\n",
       "Classes     Orders         Species                             \n",
       "Actinopteri Anabantiformes Anabas_testudineus            NaN   \n",
       "                           Betta_splendens               NaN   \n",
       "            Beloniformes   Oryzias_latipes               NaN   \n",
       "            Blenniiformes  Gouania_willdenowi            NaN   \n",
       "                           Salarias_fasciatus            NaN   \n",
       "...                                                      ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            NaN   \n",
       "                           Pelodiscus_sinensis           NaN   \n",
       "                           Pelusios_castaneus            NaN   \n",
       "                           Terrapene_carolina_triunguis  NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        NaN   \n",
       "\n",
       "                                                                         0B  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000010433   \n",
       "                           Betta_splendens               ENSBSLP00000049808   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000022912   \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000010052   \n",
       "                           Salarias_fasciatus            ENSSFAP00005052284   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            ENSGAGP00000032669   \n",
       "                           Pelodiscus_sinensis           ENSPSIP00000004165   \n",
       "                           Pelusios_castaneus            ENSPCEP00000004116   \n",
       "                           Terrapene_carolina_triunguis  ENSTMTP00000031552   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        ENSNPEP00000008724   \n",
       "\n",
       "                                                                         1A  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000011421   \n",
       "                           Betta_splendens               ENSBSLP00000036647   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000037782   \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000042079   \n",
       "                           Salarias_fasciatus            ENSSFAP00005039658   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            ENSGAGP00000002463   \n",
       "                           Pelodiscus_sinensis           ENSPSIP00000017723   \n",
       "                           Pelusios_castaneus            ENSPCEP00000016745   \n",
       "                           Terrapene_carolina_triunguis                 NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        ENSNPEP00000013660   \n",
       "\n",
       "                                                                         1B  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000026918   \n",
       "                           Betta_splendens               ENSBSLP00000013895   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000007185   \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000052479   \n",
       "                           Salarias_fasciatus            ENSSFAP00005007557   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            ENSGAGP00000015760   \n",
       "                           Pelodiscus_sinensis           ENSPSIP00000008342   \n",
       "                           Pelusios_castaneus            ENSPCEP00000021009   \n",
       "                           Terrapene_carolina_triunguis  ENSTMTP00000023774   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        ENSNPEP00000006243   \n",
       "\n",
       "                                                                         2A  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus                           NaN   \n",
       "                           Betta_splendens                              NaN   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000045928   \n",
       "            Blenniiformes  Gouania_willdenowi                           NaN   \n",
       "                           Salarias_fasciatus                           NaN   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii                           NaN   \n",
       "                           Pelodiscus_sinensis                          NaN   \n",
       "                           Pelusios_castaneus                           NaN   \n",
       "                           Terrapene_carolina_triunguis                 NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria                       NaN   \n",
       "\n",
       "                                                                         2B  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000025936   \n",
       "                           Betta_splendens               ENSBSLP00000019183   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000034755   \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000054609   \n",
       "                           Salarias_fasciatus            ENSSFAP00005037587   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii                           NaN   \n",
       "                           Pelodiscus_sinensis                          NaN   \n",
       "                           Pelusios_castaneus                           NaN   \n",
       "                           Terrapene_carolina_triunguis                 NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria                       NaN   \n",
       "\n",
       "                                                          3A   3B  \\\n",
       "Classes     Orders         Species                                  \n",
       "Actinopteri Anabantiformes Anabas_testudineus            NaN  NaN   \n",
       "                           Betta_splendens               NaN  NaN   \n",
       "            Beloniformes   Oryzias_latipes               NaN  NaN   \n",
       "            Blenniiformes  Gouania_willdenowi            NaN  NaN   \n",
       "                           Salarias_fasciatus            NaN  NaN   \n",
       "...                                                      ...  ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            NaN  NaN   \n",
       "                           Pelodiscus_sinensis           NaN  NaN   \n",
       "                           Pelusios_castaneus            NaN  NaN   \n",
       "                           Terrapene_carolina_triunguis  NaN  NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        NaN  NaN   \n",
       "\n",
       "                                                                         4A  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus                           NaN   \n",
       "                           Betta_splendens                              NaN   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000002323   \n",
       "            Blenniiformes  Gouania_willdenowi                           NaN   \n",
       "                           Salarias_fasciatus                           NaN   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            ENSGAGP00000034764   \n",
       "                           Pelodiscus_sinensis                          NaN   \n",
       "                           Pelusios_castaneus            ENSPCEP00000015691   \n",
       "                           Terrapene_carolina_triunguis  ENSTMTP00000030499   \n",
       "            Tinamiformes   Nothoprocta_perdicaria                       NaN   \n",
       "\n",
       "                                                          4B  ...  \\\n",
       "Classes     Orders         Species                            ...   \n",
       "Actinopteri Anabantiformes Anabas_testudineus            NaN  ...   \n",
       "                           Betta_splendens               NaN  ...   \n",
       "            Beloniformes   Oryzias_latipes               NaN  ...   \n",
       "            Blenniiformes  Gouania_willdenowi            NaN  ...   \n",
       "                           Salarias_fasciatus            NaN  ...   \n",
       "...                                                      ...  ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            NaN  ...   \n",
       "                           Pelodiscus_sinensis           NaN  ...   \n",
       "                           Pelusios_castaneus            NaN  ...   \n",
       "                           Terrapene_carolina_triunguis  NaN  ...   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        NaN  ...   \n",
       "\n",
       "                                                                       982A  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000031779   \n",
       "                           Betta_splendens               ENSBSLP00000000958   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000001643   \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000010024   \n",
       "                           Salarias_fasciatus            ENSSFAP00005047708   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii                           NaN   \n",
       "                           Pelodiscus_sinensis                          NaN   \n",
       "                           Pelusios_castaneus                           NaN   \n",
       "                           Terrapene_carolina_triunguis                 NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria                       NaN   \n",
       "\n",
       "                                                                       982B  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000031806   \n",
       "                           Betta_splendens               ENSBSLP00000000940   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000038989   \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000010029   \n",
       "                           Salarias_fasciatus            ENSSFAP00005047718   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            ENSGAGP00000003994   \n",
       "                           Pelodiscus_sinensis           ENSPSIP00000010788   \n",
       "                           Pelusios_castaneus            ENSPCEP00000009535   \n",
       "                           Terrapene_carolina_triunguis  ENSTMTP00000026744   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        ENSNPEP00000021207   \n",
       "\n",
       "                                                        983A 983B 984A 984B  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            NaN  NaN  NaN  NaN   \n",
       "                           Betta_splendens               NaN  NaN  NaN  NaN   \n",
       "            Beloniformes   Oryzias_latipes               NaN  NaN  NaN  NaN   \n",
       "            Blenniiformes  Gouania_willdenowi            NaN  NaN  NaN  NaN   \n",
       "                           Salarias_fasciatus            NaN  NaN  NaN  NaN   \n",
       "...                                                      ...  ...  ...  ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            NaN  NaN  NaN  NaN   \n",
       "                           Pelodiscus_sinensis           NaN  NaN  NaN  NaN   \n",
       "                           Pelusios_castaneus            NaN  NaN  NaN  NaN   \n",
       "                           Terrapene_carolina_triunguis  NaN  NaN  NaN  NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                                                       985A  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000005563   \n",
       "                           Betta_splendens               ENSBSLP00000037416   \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000013968   \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000052240   \n",
       "                           Salarias_fasciatus                           NaN   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            ENSGAGP00000004118   \n",
       "                           Pelodiscus_sinensis           ENSPSIP00000017605   \n",
       "                           Pelusios_castaneus            ENSPCEP00000015385   \n",
       "                           Terrapene_carolina_triunguis  ENSTMTP00000000781   \n",
       "            Tinamiformes   Nothoprocta_perdicaria                       NaN   \n",
       "\n",
       "                                                        985B  \\\n",
       "Classes     Orders         Species                             \n",
       "Actinopteri Anabantiformes Anabas_testudineus            NaN   \n",
       "                           Betta_splendens               NaN   \n",
       "            Beloniformes   Oryzias_latipes               NaN   \n",
       "            Blenniiformes  Gouania_willdenowi            NaN   \n",
       "                           Salarias_fasciatus            NaN   \n",
       "...                                                      ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            NaN   \n",
       "                           Pelodiscus_sinensis           NaN   \n",
       "                           Pelusios_castaneus            NaN   \n",
       "                           Terrapene_carolina_triunguis  NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria        NaN   \n",
       "\n",
       "                                                                       986A  \\\n",
       "Classes     Orders         Species                                            \n",
       "Actinopteri Anabantiformes Anabas_testudineus                           NaN   \n",
       "                           Betta_splendens                              NaN   \n",
       "            Beloniformes   Oryzias_latipes                              NaN   \n",
       "            Blenniiformes  Gouania_willdenowi                           NaN   \n",
       "                           Salarias_fasciatus                           NaN   \n",
       "...                                                                     ...   \n",
       "Sauropsida  Testudines     Gopherus_agassizii            ENSGAGP00000027308   \n",
       "                           Pelodiscus_sinensis           ENSPSIP00000004610   \n",
       "                           Pelusios_castaneus            ENSPCEP00000007761   \n",
       "                           Terrapene_carolina_triunguis                 NaN   \n",
       "            Tinamiformes   Nothoprocta_perdicaria                       NaN   \n",
       "\n",
       "                                                                       986B  \n",
       "Classes     Orders         Species                                           \n",
       "Actinopteri Anabantiformes Anabas_testudineus            ENSATEP00000014120  \n",
       "                           Betta_splendens               ENSBSLP00000026647  \n",
       "            Beloniformes   Oryzias_latipes               ENSORLP00000025592  \n",
       "            Blenniiformes  Gouania_willdenowi            ENSGWIP00000014315  \n",
       "                           Salarias_fasciatus            ENSSFAP00005047481  \n",
       "...                                                                     ...  \n",
       "Sauropsida  Testudines     Gopherus_agassizii                           NaN  \n",
       "                           Pelodiscus_sinensis                          NaN  \n",
       "                           Pelusios_castaneus                           NaN  \n",
       "                           Terrapene_carolina_triunguis                 NaN  \n",
       "            Tinamiformes   Nothoprocta_perdicaria                       NaN  \n",
       "\n",
       "[168 rows x 1974 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ort.dataframe('ENSGALP00000072350')                        # Return given accession containing dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['400A', '400B']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ort.pairs('ENSP00000256999')                            # Return pairs orthologues accession for a given ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Mammalia': [{'A': [nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan]},\n",
       "  {'B': ['ENSBBBP00000020804',\n",
       "    'ENSBTAP00000063539',\n",
       "    'ENSCDRP00005017821',\n",
       "    'ENSCHIP00000007198',\n",
       "    'ENSCWAP00000023166',\n",
       "    'ENSDLEP00000029441',\n",
       "    'ENSMMSP00000014344',\n",
       "    'ENSOARP00000011930',\n",
       "    'ENSPCTP00005016882',\n",
       "    'ENSSSCP00000046337',\n",
       "    'ENSTTRP00000009428',\n",
       "    nan,\n",
       "    'ENSAMEP00000001858',\n",
       "    'ENSCAFP00000040348',\n",
       "    'ENSFCAP00000059302',\n",
       "    'ENSLCNP00005011403',\n",
       "    'ENSMPUP00000009613',\n",
       "    'ENSNVIP00000007909',\n",
       "    'ENSPLOP00000022561',\n",
       "    'ENSSSUP00005002618',\n",
       "    'ENSUMAP00000035515',\n",
       "    'ENSVVUP00000023395',\n",
       "    'ENSMLUP00000010840',\n",
       "    'ENSPVAP00000010600',\n",
       "    'ENSRFEP00010000212',\n",
       "    'ENSDNOP00000011926',\n",
       "    'ENSSHAP00000020158',\n",
       "    'ENSMODP00000026297',\n",
       "    'ENSMEUP00000007329',\n",
       "    'ENSPCIP00000038165',\n",
       "    'ENSVURP00010030436',\n",
       "    nan,\n",
       "    nan,\n",
       "    'ENSPCAP00000009021',\n",
       "    'ENSOPRP00000004547',\n",
       "    'ENSOCUP00000034549',\n",
       "    'ENSOANP00000040484',\n",
       "    'ENSECAP00000020500',\n",
       "    nan,\n",
       "    'ENSCJAP00000029170',\n",
       "    'ENSCATP00000029508',\n",
       "    'ENSCSAP00000004297',\n",
       "    'ENSGGOP00000018048',\n",
       "    'ENSP00000381231',\n",
       "    'ENSMMUP00000065610',\n",
       "    'ENSMLEP00000035743',\n",
       "    'ENSMICP00000044325',\n",
       "    'ENSPTRP00000024013',\n",
       "    'ENSPANP00000012853',\n",
       "    'ENSPPYP00000012878',\n",
       "    'ENSTGEP00000025463',\n",
       "    'ENSLAFP00000008333',\n",
       "    'ENSCCNP00000021088',\n",
       "    'ENSCPOP00000007040',\n",
       "    'ENSCLAP00000017728',\n",
       "    'ENSDORP00000000191',\n",
       "    'ENSSTOP00000026773',\n",
       "    'ENSMUGP00000008897',\n",
       "    'ENSMAUP00000023966',\n",
       "    'ENSMOCP00000010559',\n",
       "    'ENSMUSP00000158269',\n",
       "    'ENSODEP00000017622',\n",
       "    'ENSPEMP00000025259',\n",
       "    'ENSRNOP00000039968',\n",
       "    nan]}],\n",
       " 'Sauropsida': [{'A': ['ENSANIP00000020259',\n",
       "    'ENSACCP00020011586',\n",
       "    'ENSAPLP00020000921',\n",
       "    nan,\n",
       "    'ENSAOWP00000021365',\n",
       "    'ENSDNVP00000009064',\n",
       "    'ENSCPUP00000017798',\n",
       "    'ENSCPRP00005023631',\n",
       "    nan,\n",
       "    'ENSCJPP00005009957',\n",
       "    'ENSGALP00000026056',\n",
       "    nan,\n",
       "    'ENSNMEP00000014554',\n",
       "    'ENSPSTP00000001130',\n",
       "    'ENSPCLP00000008561',\n",
       "    'ENSCPVP00000004096',\n",
       "    'ENSCCEP00000018765',\n",
       "    'ENSEGOP00005008571',\n",
       "    'ENSFALP00000002065',\n",
       "    'ENSGFOP00000021906',\n",
       "    'ENSJHYP00000018001',\n",
       "    'ENSLSDP00000022233',\n",
       "    'ENSMVIP00005009213',\n",
       "    'ENSPMJP00000014125',\n",
       "    'ENSSCAP00000015263',\n",
       "    'ENSTGUP00000005939',\n",
       "    'ENSZALP00000002179',\n",
       "    'ENSACOP00000011163',\n",
       "    'ENSMUNP00000011563',\n",
       "    'ENSSHBP00005023089',\n",
       "    nan,\n",
       "    'ENSACAP00000002508',\n",
       "    'ENSLLTP00000021252',\n",
       "    'ENSNSUP00000009778',\n",
       "    'ENSPMRP00000034240',\n",
       "    'ENSPVIP00000027376',\n",
       "    'ENSPTXP00000015326',\n",
       "    'ENSSMRP00000029804',\n",
       "    'ENSVKKP00000020727',\n",
       "    'ENSACUP00000019718',\n",
       "    'ENSSCUP00000016472',\n",
       "    'ENSCABP00000014565',\n",
       "    'ENSCSRP00000023361',\n",
       "    'ENSCPBP00000017703',\n",
       "    'ENSGAGP00000010102',\n",
       "    'ENSPSIP00000020487',\n",
       "    'ENSPCEP00000020055',\n",
       "    'ENSTMTP00000018667',\n",
       "    'ENSNPEP00000012734']},\n",
       "  {'B': ['ENSANIP00000020313',\n",
       "    'ENSACCP00020011603',\n",
       "    'ENSAPLP00020000776',\n",
       "    'ENSACDP00005007426',\n",
       "    'ENSAOWP00000021547',\n",
       "    'ENSDNVP00000009138',\n",
       "    'ENSCPUP00000017781',\n",
       "    'ENSCPRP00005023638',\n",
       "    'ENSCPIP00010000483',\n",
       "    'ENSCJPP00005009965',\n",
       "    'ENSGALP00000072350',\n",
       "    'ENSMGAP00000015449',\n",
       "    'ENSNMEP00000014585',\n",
       "    'ENSPSTP00000001225',\n",
       "    'ENSPCLP00000008539',\n",
       "    'ENSCPVP00000004298',\n",
       "    'ENSCCEP00000018762',\n",
       "    'ENSEGOP00005008582',\n",
       "    'ENSFALP00000002073',\n",
       "    'ENSGFOP00000021898',\n",
       "    'ENSJHYP00000017970',\n",
       "    'ENSLSDP00000022237',\n",
       "    'ENSMVIP00005009217',\n",
       "    'ENSPMJP00000014127',\n",
       "    'ENSSCAP00000015275',\n",
       "    'ENSTGUP00000028206',\n",
       "    'ENSZALP00000002182',\n",
       "    'ENSACOP00000011025',\n",
       "    'ENSMUNP00000011565',\n",
       "    'ENSSHBP00005000442',\n",
       "    'ENSSPUP00000023046',\n",
       "    'ENSACAP00000017559',\n",
       "    nan,\n",
       "    'ENSNSUP00000009780',\n",
       "    'ENSPMRP00000034255',\n",
       "    'ENSPVIP00000026904',\n",
       "    'ENSPTXP00000015346',\n",
       "    'ENSSMRP00000029802',\n",
       "    'ENSVKKP00000020725',\n",
       "    'ENSACUP00000019697',\n",
       "    'ENSSCUP00000016482',\n",
       "    'ENSCABP00000014566',\n",
       "    'ENSCSRP00000023213',\n",
       "    'ENSCPBP00000017708',\n",
       "    'ENSGAGP00000010113',\n",
       "    'ENSPSIP00000020507',\n",
       "    'ENSPCEP00000017166',\n",
       "    'ENSTMTP00000028167',\n",
       "    'ENSNPEP00000012752']}],\n",
       " 'Actinopteri': [{'A': [nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan,\n",
       "    nan]},\n",
       "  {'B': ['ENSATEP00000025808',\n",
       "    'ENSBSLP00000005791',\n",
       "    'ENSORLP00000043520',\n",
       "    'ENSGWIP00000028242',\n",
       "    'ENSSFAP00005044483',\n",
       "    'ENSENLP00000042797',\n",
       "    'ENSSDUP00000007083',\n",
       "    'ENSAMXP00000038555',\n",
       "    'ENSPNAP00000031038',\n",
       "    'ENSACIP00000021746',\n",
       "    'ENSACLP00000036629',\n",
       "    'ENSHBUP00000022337',\n",
       "    'ENSMZEP00005023738',\n",
       "    'ENSNBRP00000004782',\n",
       "    'ENSONIP00000002371',\n",
       "    'ENSPNYP00000030494',\n",
       "    'ENSCHAP00000001925',\n",
       "    'ENSDCDP00000032851',\n",
       "    'ENSCARP00000001513',\n",
       "    nan,\n",
       "    'ENSDARP00000053931',\n",
       "    'ENSSGRP00000093958',\n",
       "    'ENSCVAP00000012842',\n",
       "    'ENSFHEP00000031576',\n",
       "    'ENSGAFP00000027825',\n",
       "    'ENSKMAP00000000929',\n",
       "    'ENSPREP00000019222',\n",
       "    'ENSXMAP00000008170',\n",
       "    'ENSELUP00000038996',\n",
       "    'ENSGMOP00000010295',\n",
       "    'ENSNMLP00000039735',\n",
       "    'ENSPMGP00000001273',\n",
       "    'ENSEEEP00000037895',\n",
       "    'ENSMMDP00005009002',\n",
       "    'ENSSORP00005044433',\n",
       "    'ENSLBEP00000033146',\n",
       "    'ENSPKIP00000027283',\n",
       "    'ENSSFOP00015066727',\n",
       "    'ENSCGOP00000048488',\n",
       "    'ENSGACP00000019314',\n",
       "    'ENSCSEP00000001668',\n",
       "    'ENSSMAP00000022713',\n",
       "    'ENSHHUP00000008603',\n",
       "    'ENSOTSP00005012095',\n",
       "    'ENSSTUP00000077517',\n",
       "    'ENSLOCP00000001396',\n",
       "    'ENSIPUP00000020561',\n",
       "    'ENSSAUP00010048962',\n",
       "    'ENSMAMP00000032707',\n",
       "    'ENSMALP00000022611',\n",
       "    'ENSHCOP00000020193',\n",
       "    'ENSMMOP00000013420',\n",
       "    'ENSTRUP00000035065',\n",
       "    'ENSTNIP00000011140']}]}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ort.orthodict('ENSGALP00000072350', rf, False)             # Return a columns containing dictionary, False if you don't want to include nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ort.isclass('ENSGALP00000072350', 'Mammalia')              # Check the taxonomy class for a given ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Have not accessions'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ort.whereisdup('ENSP00000256999', rf, 15)             # Check in which classes a given ID is duplicated, the float is the column populousness threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# meglio non usare il linguaggio naturale"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Orthologues dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Actinopteri</th>\n",
       "      <th>Mammalia</th>\n",
       "      <th>Sauropsida</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>54</td>\n",
       "      <td>65</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "0 Actinopteri Mammalia Sauropsida\n",
       "1          54       65         49"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ts()                                                 # orders per each class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19909"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.mi('Homo_sapiens')                                   # main isoforms count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "110288"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.ai('Homo_sapiens')                                   # all isoforms count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "134"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.tc('Homo_sapiens', 'divergent')                         # duplicated genes count per duplication kind"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tandem</th>\n",
       "      <th>convergent</th>\n",
       "      <th>divergent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens</th>\n",
       "      <td>387</td>\n",
       "      <td>55</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gallus_gallus</th>\n",
       "      <td>224</td>\n",
       "      <td>30</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Danio_rerio</th>\n",
       "      <td>502</td>\n",
       "      <td>15</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tandem  convergent  divergent\n",
       "Homo_sapiens      387          55         67\n",
       "Gallus_gallus     224          30         31\n",
       "Danio_rerio       502          15         23"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.tcdf(rs, du).applymap(lambda x: int(x/2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tandem</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Total Mammalia</th>\n",
       "      <td>417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Actinopteri</th>\n",
       "      <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Sauropsida</th>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Only Mammalia</th>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Have not accessions</th>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Only Actinopteri</th>\n",
       "      <td>158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Only Sauropsida</th>\n",
       "      <td>124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mammalia and Sauropsida</th>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mammalia, Sauropsida and Actinopteri</th>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sauropsida and Actinopteri</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mammalia and Actinopteri</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Tandem\n",
       "Total Mammalia                           417\n",
       "Total Actinopteri                        301\n",
       "Total Sauropsida                         296\n",
       "Only Mammalia                            248\n",
       "Have not accessions                      246\n",
       "Only Actinopteri                         158\n",
       "Only Sauropsida                          124\n",
       "Mammalia and Sauropsida                   68\n",
       "Mammalia, Sauropsida and Actinopteri      62\n",
       "Sauropsida and Actinopteri                42\n",
       "Mammalia and Actinopteri                  39"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#stats.wherdup('Tandem', rf, 0.15)\n",
    "stats.whichdup('Tandem', rf, 0.15)                         # duplication events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>main</th>\n",
       "      <th>all</th>\n",
       "      <th>% main</th>\n",
       "      <th>tandem</th>\n",
       "      <th>convergent</th>\n",
       "      <th>divergent</th>\n",
       "      <th>% tandem</th>\n",
       "      <th>% convergent</th>\n",
       "      <th>% divergent</th>\n",
       "      <th>queries</th>\n",
       "      <th>subjects</th>\n",
       "      <th>blast ratios</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Homo_sapiens</th>\n",
       "      <td>19909</td>\n",
       "      <td>110288</td>\n",
       "      <td>18.051828</td>\n",
       "      <td>774</td>\n",
       "      <td>110</td>\n",
       "      <td>134</td>\n",
       "      <td>3.887689</td>\n",
       "      <td>0.552514</td>\n",
       "      <td>0.673062</td>\n",
       "      <td>16017</td>\n",
       "      <td>52584</td>\n",
       "      <td>3.283012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gallus_gallus</th>\n",
       "      <td>16383</td>\n",
       "      <td>28345</td>\n",
       "      <td>57.798554</td>\n",
       "      <td>448</td>\n",
       "      <td>60</td>\n",
       "      <td>62</td>\n",
       "      <td>2.734542</td>\n",
       "      <td>0.366233</td>\n",
       "      <td>0.378441</td>\n",
       "      <td>13128</td>\n",
       "      <td>42472</td>\n",
       "      <td>3.235222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Danio_rerio</th>\n",
       "      <td>23415</td>\n",
       "      <td>51919</td>\n",
       "      <td>45.099097</td>\n",
       "      <td>1004</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "      <td>4.287850</td>\n",
       "      <td>0.128123</td>\n",
       "      <td>0.196455</td>\n",
       "      <td>21475</td>\n",
       "      <td>74277</td>\n",
       "      <td>3.458766</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                main     all     % main  tandem  convergent  divergent  \\\n",
       "Homo_sapiens   19909  110288  18.051828     774         110        134   \n",
       "Gallus_gallus  16383   28345  57.798554     448          60         62   \n",
       "Danio_rerio    23415   51919  45.099097    1004          30         46   \n",
       "\n",
       "               % tandem  % convergent  % divergent  queries  subjects  \\\n",
       "Homo_sapiens   3.887689      0.552514     0.673062    16017     52584   \n",
       "Gallus_gallus  2.734542      0.366233     0.378441    13128     42472   \n",
       "Danio_rerio    4.287850      0.128123     0.196455    21475     74277   \n",
       "\n",
       "               blast ratios  \n",
       "Homo_sapiens       3.283012  \n",
       "Gallus_gallus      3.235222  \n",
       "Danio_rerio        3.458766  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aim = stats.aim(rs).reset_index()                          # Main isoforms count\n",
    "tcdf = stats.tcdf(rs, du).reset_index()                    # Duplicated genes count\n",
    "qc = stats.bc(rs).T                                        # query / subjects ratio\n",
    "qc['blast ratios'] = qc['subjects'] / qc['queries']\n",
    "\n",
    "df = stats.allinfo(aim, tcdf).join(qc)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tandem</th>\n",
       "      <th>Convergent</th>\n",
       "      <th>Divergent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Total Mammalia</th>\n",
       "      <td>417</td>\n",
       "      <td>57</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Actinopteri</th>\n",
       "      <td>301</td>\n",
       "      <td>22</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total Sauropsida</th>\n",
       "      <td>296</td>\n",
       "      <td>39</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Only Mammalia</th>\n",
       "      <td>248</td>\n",
       "      <td>33</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Have not accessions</th>\n",
       "      <td>246</td>\n",
       "      <td>9</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Only Actinopteri</th>\n",
       "      <td>158</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Only Sauropsida</th>\n",
       "      <td>124</td>\n",
       "      <td>15</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mammalia and Sauropsida</th>\n",
       "      <td>68</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mammalia, Sauropsida and Actinopteri</th>\n",
       "      <td>62</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sauropsida and Actinopteri</th>\n",
       "      <td>42</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mammalia and Actinopteri</th>\n",
       "      <td>39</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      Tandem  Convergent  Divergent\n",
       "Total Mammalia                           417          57         67\n",
       "Total Actinopteri                        301          22         28\n",
       "Total Sauropsida                         296          39         34\n",
       "Only Mammalia                            248          33         35\n",
       "Have not accessions                      246           9         24\n",
       "Only Actinopteri                         158           5          9\n",
       "Only Sauropsida                          124          15         10\n",
       "Mammalia and Sauropsida                   68          12         14\n",
       "Mammalia, Sauropsida and Actinopteri      62           7          9\n",
       "Sauropsida and Actinopteri                42           5          1\n",
       "Mammalia and Actinopteri                  39           5          9"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.alldups(0.15, rf)                                    # All duplication events"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Outofmean\n",
       "False    983\n",
       "True       4\n",
       "Name: AB, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats.outofmean('Tandem', 1000)                             # Return sequence lenghts outliers --> max-min/max"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "riprendi quello degli allineamenti (coverage)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "prova tutto con altro pesce"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "età duplicazione e orientamento duplicazione"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Uniprot features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1130"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# number of positions\n",
    "len(uni.all_scores('Divergent', 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.34"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# scores median\n",
    "np.median(np.array(uni.all_scores('Divergent', 1)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.18"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# scores percentile\n",
    "np.quantile(np.array(uni.all_scores('Tandem', 0)), 0.90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = uni.filtdf('Convergent', 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orthogroup</th>\n",
       "      <th>accession</th>\n",
       "      <th>ensembl_accession</th>\n",
       "      <th>gene</th>\n",
       "      <th>full_protein_name</th>\n",
       "      <th>full_name_ensembl</th>\n",
       "      <th>full_name_ensembl_partner</th>\n",
       "      <th>type</th>\n",
       "      <th>category</th>\n",
       "      <th>description</th>\n",
       "      <th>begin</th>\n",
       "      <th>end</th>\n",
       "      <th>position_found</th>\n",
       "      <th>score</th>\n",
       "      <th>residue_A</th>\n",
       "      <th>residue_B</th>\n",
       "      <th>alignment_A</th>\n",
       "      <th>alignment_B</th>\n",
       "      <th>EC_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>228</th>\n",
       "      <td>12A</td>\n",
       "      <td>Q8NI17</td>\n",
       "      <td>ENSP00000498630</td>\n",
       "      <td>IL31RA</td>\n",
       "      <td>Interleukin-31 receptor subunit alpha</td>\n",
       "      <td>interleukin 31 receptor A</td>\n",
       "      <td>interleukin 6 signal transducer</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>No effect on STAT1 and STAT3 activation. Sligh...</td>\n",
       "      <td>670.0</td>\n",
       "      <td>670.0</td>\n",
       "      <td>671.0</td>\n",
       "      <td>1.10</td>\n",
       "      <td>Y</td>\n",
       "      <td>S</td>\n",
       "      <td>Y-YYYYYYYY-YYYYYYYYYYY-YYHYYYY-YSYYYYYYYYYLYYY...</td>\n",
       "      <td>SSSSSSSSS--SSSSSSSXSSS-SS-SCSCSSSSSSSSSSSSSSSS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>229</th>\n",
       "      <td>12A</td>\n",
       "      <td>Q8NI17</td>\n",
       "      <td>ENSP00000498630</td>\n",
       "      <td>IL31RA</td>\n",
       "      <td>Interleukin-31 receptor subunit alpha</td>\n",
       "      <td>interleukin 31 receptor A</td>\n",
       "      <td>interleukin 6 signal transducer</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>No effect on STAT3 and STAT5 activation. Mild ...</td>\n",
       "      <td>670.0</td>\n",
       "      <td>670.0</td>\n",
       "      <td>671.0</td>\n",
       "      <td>1.10</td>\n",
       "      <td>Y</td>\n",
       "      <td>S</td>\n",
       "      <td>Y-YYYYYYYY-YYYYYYYYYYY-YYHYYYY-YSYYYYYYYYYLYYY...</td>\n",
       "      <td>SSSSSSSSS--SSSSSSSXSSS-SS-SCSCSSSSSSSSSSSSSSSS...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>12B</td>\n",
       "      <td>P40189</td>\n",
       "      <td>ENSP00000370698</td>\n",
       "      <td>IL6ST</td>\n",
       "      <td>Interleukin-6 receptor subunit beta</td>\n",
       "      <td>interleukin 6 signal transducer</td>\n",
       "      <td>interleukin 31 receptor A</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>Induces ligand-independent activation.</td>\n",
       "      <td>172.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>171.0</td>\n",
       "      <td>1.39</td>\n",
       "      <td>M</td>\n",
       "      <td>I</td>\n",
       "      <td>MMMMMMMMMMMMMMMMMMMMMMMMMMSMMMSMMMMMMMMMMMSMMT...</td>\n",
       "      <td>IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII-IIIII...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>16A</td>\n",
       "      <td>Q16552</td>\n",
       "      <td>ENSP00000497968</td>\n",
       "      <td>IL17A</td>\n",
       "      <td>Interleukin-17A</td>\n",
       "      <td>interleukin 17A</td>\n",
       "      <td>interleukin 17F</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>Decreases the affinity for IL17RA by 5-fold.</td>\n",
       "      <td>78.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>1.38</td>\n",
       "      <td>R</td>\n",
       "      <td>X</td>\n",
       "      <td>RRPRRRRQRRRRPRRLRRRRRRRRRRRRRRRRRRRPPRRRRRRRRR...</td>\n",
       "      <td>VIIIIIIIIIIVIVIFIIIIIIXVIIIIVIVIIIVIIIIIEVNIII...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>16A</td>\n",
       "      <td>Q16552</td>\n",
       "      <td>ENSP00000497968</td>\n",
       "      <td>IL17A</td>\n",
       "      <td>Interleukin-17A</td>\n",
       "      <td>interleukin 17A</td>\n",
       "      <td>interleukin 17F</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>Decreases the affinity for IL17RA by 5-fold.</td>\n",
       "      <td>78.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>1.69</td>\n",
       "      <td>N</td>\n",
       "      <td>T</td>\n",
       "      <td>NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTNNNNNN...</td>\n",
       "      <td>TTTTTTTTTTSTTTTMTTTTTTXSTTTTTTTTTTTTTTTTTTSTTT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>16B</td>\n",
       "      <td>Q96PD4</td>\n",
       "      <td>ENSP00000337432</td>\n",
       "      <td>IL17F</td>\n",
       "      <td>Interleukin-17F</td>\n",
       "      <td>interleukin 17F</td>\n",
       "      <td>interleukin 17A</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>Significantly decreases the affinity for IL17R...</td>\n",
       "      <td>77.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>1.38</td>\n",
       "      <td>R</td>\n",
       "      <td>X</td>\n",
       "      <td>RRPRRRRQRRRRPRRLRRRRRRRRRRRRRRRRRRRPPRRRRRRRRR...</td>\n",
       "      <td>VIIIIIIIIIIVIVIFIIIIIIXVIIIIVIVIIIVIIIIIEVNIII...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>725</th>\n",
       "      <td>48B</td>\n",
       "      <td>P27338</td>\n",
       "      <td>ENSP00000367309</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>Amine oxidase [flavin-containing] B</td>\n",
       "      <td>monoamine oxidase B</td>\n",
       "      <td>monoamine oxidase A</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>No loss of activity.</td>\n",
       "      <td>172.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>173.0</td>\n",
       "      <td>1.02</td>\n",
       "      <td>F</td>\n",
       "      <td>L</td>\n",
       "      <td>FFFFFFFFFFFFFXXFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...</td>\n",
       "      <td>LLLXLLLLLFLILLLLLLLLLLLIXL-LLLLLLLLLLLLLLLELLL...</td>\n",
       "      <td>1.4.3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>726</th>\n",
       "      <td>48B</td>\n",
       "      <td>P27338</td>\n",
       "      <td>ENSP00000367309</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>Amine oxidase [flavin-containing] B</td>\n",
       "      <td>monoamine oxidase B</td>\n",
       "      <td>monoamine oxidase A</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>No loss of activity.</td>\n",
       "      <td>297.0</td>\n",
       "      <td>297.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>1.31</td>\n",
       "      <td>Q</td>\n",
       "      <td>T</td>\n",
       "      <td>QQQQQQQQ-QQXQQQQQQQQQQQQIQQQQQQQQQQQQQQQQQ-QQQ...</td>\n",
       "      <td>TTTSTTTTTSSTTTSSGTTTTNTSXT-TTTTTSTTTTTTTTTQTTS...</td>\n",
       "      <td>1.4.3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>48B</td>\n",
       "      <td>P27338</td>\n",
       "      <td>ENSP00000367309</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>Amine oxidase [flavin-containing] B</td>\n",
       "      <td>monoamine oxidase B</td>\n",
       "      <td>monoamine oxidase A</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>No loss of activity.</td>\n",
       "      <td>312.0</td>\n",
       "      <td>312.0</td>\n",
       "      <td>313.0</td>\n",
       "      <td>1.10</td>\n",
       "      <td>A</td>\n",
       "      <td>P</td>\n",
       "      <td>AAAAAAAA-AAXAAAAAAAAAAAAXAAAAAAAAAAAAAAAAA-SAA...</td>\n",
       "      <td>PPPXPPPPPPPPPPPPAPPPPPPPXP-PPPPPPPPPPPPPPPAPPP...</td>\n",
       "      <td>1.4.3.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728</th>\n",
       "      <td>48B</td>\n",
       "      <td>P27338</td>\n",
       "      <td>ENSP00000367309</td>\n",
       "      <td>MAOB</td>\n",
       "      <td>Amine oxidase [flavin-containing] B</td>\n",
       "      <td>monoamine oxidase B</td>\n",
       "      <td>monoamine oxidase A</td>\n",
       "      <td>MUTAGEN</td>\n",
       "      <td>MUTAGENESIS</td>\n",
       "      <td>Complete loss of activity.</td>\n",
       "      <td>365.0</td>\n",
       "      <td>365.0</td>\n",
       "      <td>365.0</td>\n",
       "      <td>1.66</td>\n",
       "      <td>H</td>\n",
       "      <td>T</td>\n",
       "      <td>HTNHXHSH-HQHHHHHHQHHHHHAQHHHHHHHHSHHHHHHHH-THX...</td>\n",
       "      <td>TLTTTTTTTTTTTTTTTTTTTXTTTT-TTTTTSTTTTTT-TTDTKT...</td>\n",
       "      <td>1.4.3.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    orthogroup accession ensembl_accession    gene  \\\n",
       "228        12A    Q8NI17   ENSP00000498630  IL31RA   \n",
       "229        12A    Q8NI17   ENSP00000498630  IL31RA   \n",
       "261        12B    P40189   ENSP00000370698   IL6ST   \n",
       "416        16A    Q16552   ENSP00000497968   IL17A   \n",
       "417        16A    Q16552   ENSP00000497968   IL17A   \n",
       "426        16B    Q96PD4   ENSP00000337432   IL17F   \n",
       "725        48B    P27338   ENSP00000367309    MAOB   \n",
       "726        48B    P27338   ENSP00000367309    MAOB   \n",
       "727        48B    P27338   ENSP00000367309    MAOB   \n",
       "728        48B    P27338   ENSP00000367309    MAOB   \n",
       "\n",
       "                         full_protein_name                full_name_ensembl  \\\n",
       "228  Interleukin-31 receptor subunit alpha        interleukin 31 receptor A   \n",
       "229  Interleukin-31 receptor subunit alpha        interleukin 31 receptor A   \n",
       "261    Interleukin-6 receptor subunit beta  interleukin 6 signal transducer   \n",
       "416                        Interleukin-17A                  interleukin 17A   \n",
       "417                        Interleukin-17A                  interleukin 17A   \n",
       "426                        Interleukin-17F                  interleukin 17F   \n",
       "725    Amine oxidase [flavin-containing] B              monoamine oxidase B   \n",
       "726    Amine oxidase [flavin-containing] B              monoamine oxidase B   \n",
       "727    Amine oxidase [flavin-containing] B              monoamine oxidase B   \n",
       "728    Amine oxidase [flavin-containing] B              monoamine oxidase B   \n",
       "\n",
       "           full_name_ensembl_partner     type     category  \\\n",
       "228  interleukin 6 signal transducer  MUTAGEN  MUTAGENESIS   \n",
       "229  interleukin 6 signal transducer  MUTAGEN  MUTAGENESIS   \n",
       "261        interleukin 31 receptor A  MUTAGEN  MUTAGENESIS   \n",
       "416                  interleukin 17F  MUTAGEN  MUTAGENESIS   \n",
       "417                  interleukin 17F  MUTAGEN  MUTAGENESIS   \n",
       "426                  interleukin 17A  MUTAGEN  MUTAGENESIS   \n",
       "725              monoamine oxidase A  MUTAGEN  MUTAGENESIS   \n",
       "726              monoamine oxidase A  MUTAGEN  MUTAGENESIS   \n",
       "727              monoamine oxidase A  MUTAGEN  MUTAGENESIS   \n",
       "728              monoamine oxidase A  MUTAGEN  MUTAGENESIS   \n",
       "\n",
       "                                           description  begin    end  \\\n",
       "228  No effect on STAT1 and STAT3 activation. Sligh...  670.0  670.0   \n",
       "229  No effect on STAT3 and STAT5 activation. Mild ...  670.0  670.0   \n",
       "261             Induces ligand-independent activation.  172.0  172.0   \n",
       "416       Decreases the affinity for IL17RA by 5-fold.   78.0   78.0   \n",
       "417       Decreases the affinity for IL17RA by 5-fold.   78.0   78.0   \n",
       "426  Significantly decreases the affinity for IL17R...   77.0   77.0   \n",
       "725                               No loss of activity.  172.0  172.0   \n",
       "726                               No loss of activity.  297.0  297.0   \n",
       "727                               No loss of activity.  312.0  312.0   \n",
       "728                         Complete loss of activity.  365.0  365.0   \n",
       "\n",
       "     position_found  score residue_A residue_B  \\\n",
       "228           671.0   1.10         Y         S   \n",
       "229           671.0   1.10         Y         S   \n",
       "261           171.0   1.39         M         I   \n",
       "416            78.0   1.38         R         X   \n",
       "417            79.0   1.69         N         T   \n",
       "426            78.0   1.38         R         X   \n",
       "725           173.0   1.02         F         L   \n",
       "726           296.0   1.31         Q         T   \n",
       "727           313.0   1.10         A         P   \n",
       "728           365.0   1.66         H         T   \n",
       "\n",
       "                                           alignment_A  \\\n",
       "228  Y-YYYYYYYY-YYYYYYYYYYY-YYHYYYY-YSYYYYYYYYYLYYY...   \n",
       "229  Y-YYYYYYYY-YYYYYYYYYYY-YYHYYYY-YSYYYYYYYYYLYYY...   \n",
       "261  MMMMMMMMMMMMMMMMMMMMMMMMMMSMMMSMMMMMMMMMMMSMMT...   \n",
       "416  RRPRRRRQRRRRPRRLRRRRRRRRRRRRRRRRRRRPPRRRRRRRRR...   \n",
       "417  NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNTNNNNNN...   \n",
       "426  RRPRRRRQRRRRPRRLRRRRRRRRRRRRRRRRRRRPPRRRRRRRRR...   \n",
       "725  FFFFFFFFFFFFFXXFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF...   \n",
       "726  QQQQQQQQ-QQXQQQQQQQQQQQQIQQQQQQQQQQQQQQQQQ-QQQ...   \n",
       "727  AAAAAAAA-AAXAAAAAAAAAAAAXAAAAAAAAAAAAAAAAA-SAA...   \n",
       "728  HTNHXHSH-HQHHHHHHQHHHHHAQHHHHHHHHSHHHHHHHH-THX...   \n",
       "\n",
       "                                           alignment_B EC_number  \n",
       "228  SSSSSSSSS--SSSSSSSXSSS-SS-SCSCSSSSSSSSSSSSSSSS...       NaN  \n",
       "229  SSSSSSSSS--SSSSSSSXSSS-SS-SCSCSSSSSSSSSSSSSSSS...       NaN  \n",
       "261  IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII-IIIII...       NaN  \n",
       "416  VIIIIIIIIIIVIVIFIIIIIIXVIIIIVIVIIIVIIIIIEVNIII...       NaN  \n",
       "417  TTTTTTTTTTSTTTTMTTTTTTXSTTTTTTTTTTTTTTTTTTSTTT...       NaN  \n",
       "426  VIIIIIIIIIIVIVIFIIIIIIXVIIIIVIVIIIVIIIIIEVNIII...       NaN  \n",
       "725  LLLXLLLLLFLILLLLLLLLLLLIXL-LLLLLLLLLLLLLLLELLL...   1.4.3.4  \n",
       "726  TTTSTTTTTSSTTTSSGTTTTNTSXT-TTTTTSTTTTTTTTTQTTS...   1.4.3.4  \n",
       "727  PPPXPPPPPPPPPPPPAPPPPPPPXP-PPPPPPPPPPPPPPPAPPP...   1.4.3.4  \n",
       "728  TLTTTTTTTTTTTTTTTTTTTXTTTT-TTTTTSTTTTTT-TTDTKT...   1.4.3.4  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['type'] == 'MUTAGEN'].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tandem</th>\n",
       "      <th>Unique_tandem</th>\n",
       "      <th>Convergent</th>\n",
       "      <th>Unique_convergent</th>\n",
       "      <th>Divergent</th>\n",
       "      <th>Unique_divergent</th>\n",
       "      <th>Sum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACT_SITE</th>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BINDING</th>\n",
       "      <td>16.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CARBOHYD</th>\n",
       "      <td>150.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>191.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CA_BIND</th>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHAIN</th>\n",
       "      <td>83.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COILED</th>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COMPBIAS</th>\n",
       "      <td>23.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>32.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CONFLICT</th>\n",
       "      <td>133.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>157.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CROSSLNK</th>\n",
       "      <td>23.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DISULFID</th>\n",
       "      <td>164.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>243.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNA_BIND</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOMAIN</th>\n",
       "      <td>156.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>228.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INIT_MET</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIPID</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>METAL</th>\n",
       "      <td>37.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOD_RES</th>\n",
       "      <td>92.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>138.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOTIF</th>\n",
       "      <td>18.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MUTAGEN</th>\n",
       "      <td>101.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>117.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NON_TER</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_BIND</th>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROPEP</th>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REGION</th>\n",
       "      <td>78.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>97.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REPEAT</th>\n",
       "      <td>101.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>122.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIGNAL</th>\n",
       "      <td>18.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SITE</th>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VARIANT</th>\n",
       "      <td>328.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>430.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VAR_SEQ</th>\n",
       "      <td>98.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>165.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN_FING</th>\n",
       "      <td>70.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>82.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PEPTIDE</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRANSIT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Tandem  Unique_tandem  Convergent  Unique_convergent  Divergent  \\\n",
       "ACT_SITE    10.0           10.0         0.0                0.0        2.0   \n",
       "BINDING     16.0           11.0         1.0                1.0        6.0   \n",
       "CARBOHYD   150.0           83.0        32.0               17.0        9.0   \n",
       "CA_BIND      4.0            2.0         0.0                0.0        1.0   \n",
       "CHAIN       83.0           63.0         9.0                9.0       16.0   \n",
       "COILED       6.0            4.0         0.0                0.0        1.0   \n",
       "COMPBIAS    23.0           14.0         3.0                3.0        6.0   \n",
       "CONFLICT   133.0           70.0         6.0                6.0       18.0   \n",
       "CROSSLNK    23.0            8.0         0.0                0.0       13.0   \n",
       "DISULFID   164.0           83.0        76.0               22.0        3.0   \n",
       "DNA_BIND     2.0            1.0         0.0                0.0        3.0   \n",
       "DOMAIN     156.0           76.0        43.0               13.0       29.0   \n",
       "INIT_MET     1.0            1.0         0.0                0.0        2.0   \n",
       "LIPID        3.0            2.0         0.0                0.0        0.0   \n",
       "METAL       37.0           22.0         2.0                2.0        8.0   \n",
       "MOD_RES     92.0           41.0        17.0                6.0       29.0   \n",
       "MOTIF       18.0            8.0         2.0                2.0        3.0   \n",
       "MUTAGEN    101.0           42.0        10.0                5.0        6.0   \n",
       "NON_TER      1.0            1.0         0.0                0.0        0.0   \n",
       "NP_BIND      4.0            4.0         0.0                0.0        1.0   \n",
       "PROPEP       8.0            7.0         7.0                4.0        1.0   \n",
       "REGION      78.0           36.0         9.0                5.0       10.0   \n",
       "REPEAT     101.0           28.0        11.0                5.0       10.0   \n",
       "SIGNAL      18.0           12.0         4.0                4.0        1.0   \n",
       "SITE        17.0            9.0         3.0                2.0        1.0   \n",
       "VARIANT    328.0          126.0        40.0               25.0       62.0   \n",
       "VAR_SEQ     98.0           52.0        35.0               13.0       32.0   \n",
       "ZN_FING     70.0           20.0         0.0                0.0       12.0   \n",
       "PEPTIDE      0.0            0.0         1.0                1.0        0.0   \n",
       "TRANSIT      0.0            0.0         0.0                0.0        1.0   \n",
       "\n",
       "          Unique_divergent    Sum  \n",
       "ACT_SITE               2.0   12.0  \n",
       "BINDING                4.0   23.0  \n",
       "CARBOHYD               6.0  191.0  \n",
       "CA_BIND                1.0    5.0  \n",
       "CHAIN                 13.0  108.0  \n",
       "COILED                 1.0    7.0  \n",
       "COMPBIAS               3.0   32.0  \n",
       "CONFLICT               9.0  157.0  \n",
       "CROSSLNK               6.0   36.0  \n",
       "DISULFID               2.0  243.0  \n",
       "DNA_BIND               2.0    5.0  \n",
       "DOMAIN                14.0  228.0  \n",
       "INIT_MET               2.0    3.0  \n",
       "LIPID                  0.0    3.0  \n",
       "METAL                  4.0   47.0  \n",
       "MOD_RES               12.0  138.0  \n",
       "MOTIF                  2.0   23.0  \n",
       "MUTAGEN                3.0  117.0  \n",
       "NON_TER                0.0    1.0  \n",
       "NP_BIND                1.0    5.0  \n",
       "PROPEP                 1.0   16.0  \n",
       "REGION                 6.0   97.0  \n",
       "REPEAT                 5.0  122.0  \n",
       "SIGNAL                 1.0   23.0  \n",
       "SITE                   1.0   21.0  \n",
       "VARIANT               22.0  430.0  \n",
       "VAR_SEQ               17.0  165.0  \n",
       "ZN_FING                6.0   82.0  \n",
       "PEPTIDE                0.0    1.0  \n",
       "TRANSIT                1.0    1.0  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# all features count\n",
    "df = uni.allfeat(None).fillna(0)\n",
    "# mettere niente come argomento per il default?\n",
    "# aggiungere argomento range (+1-1)\n",
    "df['Sum'] = df['Tandem'] + df['Convergent'] + df['Divergent']\n",
    "df.astype(int)\n",
    "# prova tabella finale con \"duplication\" e \"unique gene\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ort:\n",
    "    \n",
    "    def whichdup(acc):\n",
    "\n",
    "        def read(kind):\n",
    "            path = cwd + '/orthologues/' + kind + '_orthologues.csv'\n",
    "            return pd.read_table(path, sep=';')\n",
    "\n",
    "        t = read('Tandem').isin([acc]).any()\n",
    "        d = read('Divergent').isin([acc]).any()\n",
    "        c = read('Convergent').isin([acc]).any()\n",
    "\n",
    "        if any(t):\n",
    "            return 'Tandem'\n",
    "        elif any(d):\n",
    "            return 'Divergent'\n",
    "        elif any(c):\n",
    "            return 'Convergent'\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "        #ort.whichdup('ENSGALP00000072350')\n",
    "        \n",
    "    def dataframe(acc):\n",
    "        kind = ort.whichdup(acc)\n",
    "        path = cwd + '/orthologues/' + kind + '_orthologues.csv'\n",
    "        df = pd.read_table(path, sep=';')\n",
    "        return df.set_index(['Classes', 'Orders', 'Species'])\n",
    "        \n",
    "        #ort.dataframe('ENSGALP00000072350')\n",
    "\n",
    "    def pairs(acc):\n",
    "        df = ort.dataframe(acc)\n",
    "        p = df.columns[df.isin([acc]).any()][0]\n",
    "        p_num = re.split('A|B', p)[0]\n",
    "\n",
    "        if 'B' in p:\n",
    "            pairs = [p_num + 'A', p]\n",
    "        else:\n",
    "            pairs = [p, p_num + 'B']\n",
    "\n",
    "        return pairs\n",
    "\n",
    "        #ort.pairs('ENSGALP00000072350')\n",
    "\n",
    "    def ortholist(acc, ref_classes, includenone):\n",
    "        df = ort.dataframe(acc)\n",
    "        df = df.loc[ref_classes]\n",
    "        df = df[ort.pairs(acc)].values.tolist()\n",
    "\n",
    "        if not includenone:\n",
    "            dfA = [l[0] for l in df \n",
    "                   if str(l[0]) != 'nan']\n",
    "            dfB = [l[1] for l in df \n",
    "                   if str(l[1]) != 'nan']\n",
    "        else:\n",
    "            dfA = [l[0] for l in df]\n",
    "            dfB = [l[1] for l in df]\n",
    "            \n",
    "        return [dfA, dfB]\n",
    "\n",
    "        #ort.ortholist('ENSGALP00000072350', 'Sauropsida', False)\n",
    "\n",
    "    def orthodict(acc, ref_classes, includenone):\n",
    "        df = ort.dataframe(acc)\n",
    "        c = {}\n",
    "        for line in ref_classes:\n",
    "            df2 = df[ort.pairs(acc)].loc[line].values.tolist()\n",
    "\n",
    "            if not includenone:\n",
    "                dfA = [l[0] for l in df2 \n",
    "                       if str(l[0]) != 'nan']\n",
    "                dfB = [l[1] for l in df2 \n",
    "                       if str(l[1]) != 'nan']\n",
    "            else:\n",
    "                dfA = [l[0] for l in df2]\n",
    "                dfB = [l[1] for l in df2]\n",
    "            c.update({line: [{'A': dfA}, {'B': dfB}]})\n",
    "        return c\n",
    "    \n",
    "        #ort.orthodict('ENSGALP00000072350', rf, True)\n",
    "\n",
    "    def isclass(acc, cla):\n",
    "        df = ort.dataframe(acc)\n",
    "        df = df[ort.pairs(acc)].loc[cla]\n",
    "        numna = len(df.values.tolist())\n",
    "        num = len(df.iloc[:,0].dropna().tolist())\n",
    "        if num > numna * 0.15:\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "\n",
    "        #ort.isclass('ENSGALP00000072350', 'Sauropsida')\n",
    "        \n",
    "    def whereisdup(ac, rf, t):\n",
    "\n",
    "        path = cwd + '/orthologues/' + ort.whichdup(ac) + '_orthologues.csv'\n",
    "        df = pd.read_table(path, sep=';')\n",
    "        df = df.set_index(['Classes', 'Orders', 'Species'])\n",
    "\n",
    "        m = df[ort.pairs(ac)].loc[rf[0]].count()\n",
    "        s = df[ort.pairs(ac)].loc[rf[1]].count()\n",
    "        a = df[ort.pairs(ac)].loc[rf[2]].count()\n",
    "\n",
    "        tsa = len(df.loc[rf[0]].index.tolist())\n",
    "        tsm = len(df.loc[rf[1]].index.tolist())\n",
    "        tss = len(df.loc[rf[2]].index.tolist())\n",
    "\n",
    "        if m.tolist()[0] > tsm*t and m.tolist()[1] > tsm*t:\n",
    "            ma = rf[0]\n",
    "        else:\n",
    "            ma = None\n",
    "\n",
    "        if s.tolist()[0] > tss*t and s.tolist()[1] > tss*t:\n",
    "            sa = rf[1]\n",
    "        else:\n",
    "            sa = None\n",
    "\n",
    "        if a.tolist()[0] > tsa*t and a.tolist()[1] > tsa*t:\n",
    "            ac = rf[2]\n",
    "        else:\n",
    "            ac = None\n",
    "\n",
    "        mat = {\n",
    "                str(['Mammalia', None, None]): 'Only Mammalia',\n",
    "                str([None, 'Sauropsida', None]): 'Only Sauropsida',\n",
    "                str([None, None, 'Actinopteri']): 'Only Actinopteri',\n",
    "                str(['Mammalia', 'Sauropsida', None]): 'Mammalia and Sauropsida',\n",
    "                str(['Mammalia', None, 'Actinopteri']): 'Mammalia and Actinopteri',\n",
    "                str([None, 'Sauropsida', 'Actinopteri']): 'Sauropsida and Actinopteri',\n",
    "                str(['Mammalia', 'Sauropsida', 'Actinopteri']): 'Mammalia, Sauropsida and Actinopteri',\n",
    "                str([None, None, None]): 'Have not accessions',\n",
    "                }\n",
    "\n",
    "        return mat.get(str([ma, sa, ac]))\n",
    "\n",
    "        #ort.whereisdup('ENSGALP00000072350', rf, 0.15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class stats():\n",
    "\n",
    "    # species classes count\n",
    "    def ts():\n",
    "        path = 'species_list.txt'\n",
    "        s = pd.read_table(path, sep=' ', header=None)\n",
    "        df = pd.DataFrame(s.groupby(0).count()[1]).reset_index().T\n",
    "        df.columns = df.iloc[0]\n",
    "        return df[1:]\n",
    "\n",
    "        # stats.ts()\n",
    "\n",
    "    # all isoforms count\n",
    "    def ai(s):\n",
    "        a = []\n",
    "        path = cwd + '/fa/' + s + '.fa.gz'\n",
    "        fa = SeqIO.parse(gzip.open(path, 'rt'), 'fasta')\n",
    "        for fasta in fa:\n",
    "            if 'gene_biotype:protein_coding' in fasta.description:\n",
    "                a.append(fasta.id)\n",
    "        return len(a)\n",
    "\n",
    "        # stats.ai('Homo_sapiens')\n",
    "\n",
    "    # main isoforms count\n",
    "    def mi(s):\n",
    "        path = cwd + '/main/' + s + '.tsv'\n",
    "        df = pd.read_table(path, sep='\\t')\n",
    "        count = df['Protein_id'].count().tolist()\n",
    "        return count\n",
    "\n",
    "        # stats.mi('Homo_sapiens')\n",
    "        \n",
    "    # all and main isoforms ratio\n",
    "    def aim(rs):\n",
    "        mi = [stats.mi(x) for x in rs]\n",
    "        ai = [stats.ai(x) for x in rs]\n",
    "        index=['main', 'all']\n",
    "        df = pd.DataFrame([mi, ai], index=index, columns=rs).T\n",
    "        df['% main'] = df['main'] * 100 / df['all']\n",
    "        return df\n",
    "    \n",
    "        # stats.aim('Homo_sapiens')\n",
    "\n",
    "    # duplicated genes count\n",
    "    def tc(s, k):\n",
    "        path = cwd + '/tandem/' + s + '_' + k + '.tsv'\n",
    "        df = pd.read_table(path, sep='\\t', header=None)\n",
    "        df = df.values.tolist()\n",
    "        return len(df)*2\n",
    "    \n",
    "        # stats.tc('Homo_sapiens', 'tandem')\n",
    "        \n",
    "    def tcdf(rs, du):\n",
    "        tcdf = {\n",
    "                rs[0]: {\n",
    "                    du[0]: stats.tc(rs[0], du[0]), \n",
    "                    du[1]: stats.tc(rs[0], du[1]), \n",
    "                    du[2]: stats.tc(rs[0], du[2]),\n",
    "                    },\n",
    "                rs[1]: {\n",
    "                    du[0]: stats.tc(rs[1], du[0]), \n",
    "                    du[1]: stats.tc(rs[1], du[1]), \n",
    "                    du[2]: stats.tc(rs[1], du[2]),\n",
    "                    },\n",
    "                rs[2]: {\n",
    "                    du[0]: stats.tc(rs[2], du[0]), \n",
    "                    du[1]: stats.tc(rs[2], du[1]), \n",
    "                    du[2]: stats.tc(rs[2], du[2]),\n",
    "                    },\n",
    "        }\n",
    "\n",
    "        return pd.DataFrame(tcdf).T\n",
    "    \n",
    "        #stats.tcdf(rs, du)\n",
    "        \n",
    "    # return numbers and percentage about main species\n",
    "    def allinfo(df1, df2):\n",
    "        df = pd.merge(df1, df2, on='index')\n",
    "        df['% tandem'] = df['tandem'] * 100 / df['main']\n",
    "        df['% convergent'] = df['convergent'] * 100 / df['main']\n",
    "        df['% divergent'] = df['divergent'] * 100 / df['main']\n",
    "        df = df.set_index('index')\n",
    "        df.index.name = None\n",
    "        \n",
    "        return df\n",
    "    \n",
    "        # stats.allinfo(stats.aim(rs).reset_index(), stats.tcdf(rs, du).reset_index())\n",
    "\n",
    "    # blast comparisons count\n",
    "    def bcf(specie):\n",
    "        path = cwd + '/blast_queries/' + specie + '_main_blast.txt'\n",
    "        x = pd.read_table(path, comment='#', header=None)\n",
    "        x = x[x[2] != 100]\n",
    "        q = x.drop_duplicates(0)[0].count().tolist()\n",
    "        s = x[1].count().tolist()\n",
    "        return [q, s]\n",
    "    \n",
    "    def bc(rs):\n",
    "        df = pd.DataFrame({\n",
    "        rs[0]: {'queries': stats.bcf(rs[0])[0], \n",
    "                'subjects': stats.bcf(rs[0])[1]},\n",
    "        rs[1]: {'queries': stats.bcf(rs[1])[0], \n",
    "                'subjects': stats.bcf(rs[1])[1]},\n",
    "        rs[2]: {'queries': stats.bcf(rs[2])[0], \n",
    "                'subjects': stats.bcf(rs[2])[1]}})\n",
    "        return df\n",
    "\n",
    "        #stats.bc(rs)\n",
    "        \n",
    "    def wherdup(kind, rf, threshold):\n",
    "\n",
    "        def isdup(kind, c):\n",
    "            path = cwd + '/orthologues/' + kind + '_orthologues.csv'\n",
    "            df = pd.read_table(path, sep=';')\n",
    "            df = df.set_index(['Classes', 'Orders', 'Species'])\n",
    "\n",
    "            # total number of accessions\n",
    "            it = itertools.chain.from_iterable(df.values.tolist())\n",
    "            accessions = len([l for l in list(it) if not str(l) == 'nan'])\n",
    "\n",
    "            # if both columns sum is > 20%\n",
    "            df = df.applymap(lambda x: 1 if str(x) != 'nan' else 0).loc[c]\n",
    "            df = df.reset_index().drop(columns=['Orders', 'Species']).T\n",
    "            ts = int(df.columns.tolist()[-1])\n",
    "            df['sum'] = df.sum(axis=1)\n",
    "            df['Isdup'] = df['sum'].apply(lambda x: 1 if x > ts*threshold else 0) \n",
    "\n",
    "            def pairwise(iterable):\n",
    "                \"s -> (s0, s1), (s2, s3), (s4, s5), ...\"\n",
    "                a = iter(iterable)\n",
    "                return zip(a, a)\n",
    "            lp = list(pairwise(df.index))\n",
    "\n",
    "            x = pd.DataFrame([df.loc[list(l)]['Isdup'].tolist() for l in lp])\n",
    "            x['Isdup_' + c] = x.sum(axis=1)  \n",
    "            x = x['Isdup_' + c].reset_index()\n",
    "\n",
    "            return x\n",
    "\n",
    "            #stats.isdup('Convergent', 'Actinopteri')\n",
    "\n",
    "        df_list = []\n",
    "        for x in range(len(rf)):\n",
    "            df_list.append(isdup(kind, rf[x]))\n",
    "\n",
    "        msa = reduce(lambda df1,df2: pd.merge(df1,df2,on='index'), df_list)\n",
    "        msa = msa.drop(columns='index')\n",
    "\n",
    "        return msa\n",
    "    \n",
    "        # stats.wherdup('Tandem', rf, 0.15)\n",
    "\n",
    "    def whichdup(kind, rf, threshold):\n",
    "\n",
    "        df = stats.wherdup(kind, rf, threshold)\n",
    "        df['m'] = df['Isdup_Mammalia'].apply(lambda x: 'Mammalia' if x == 2 else None)\n",
    "        df['s'] = df['Isdup_Sauropsida'].apply(lambda x: 'Sauropsida' if x == 2 else None)\n",
    "        df['a'] = df['Isdup_Actinopteri'].apply(lambda x: 'Actinopteri' if x == 2 else None)\n",
    "        df = df[['m', 's', 'a']]\n",
    "\n",
    "        mat = {\n",
    "                    str(['Mammalia', None, None]): 'Only Mammalia',\n",
    "                    str([None, 'Sauropsida', None]): 'Only Sauropsida',\n",
    "                    str([None, None, 'Actinopteri']): 'Only Actinopteri',\n",
    "                    str(['Mammalia', 'Sauropsida', None]): 'Mammalia and Sauropsida',\n",
    "                    str(['Mammalia', None, 'Actinopteri']): 'Mammalia and Actinopteri',\n",
    "                    str([None, 'Sauropsida', 'Actinopteri']): 'Sauropsida and Actinopteri',\n",
    "                    str(['Mammalia', 'Sauropsida', 'Actinopteri']): 'Mammalia, Sauropsida and Actinopteri',\n",
    "                    str([None, None, None]): 'Have not accessions',\n",
    "        }\n",
    "\n",
    "        df['msa'] = df.values.tolist()\n",
    "        df['which'] = df['msa'].apply(lambda x: mat.get(str(x)))\n",
    "        df = pd.DataFrame(df['which'].value_counts().to_dict(), index=[kind])\n",
    "\n",
    "\n",
    "        def addsum(df, k):\n",
    "            classes = [l for l in df.columns.tolist() if k in l]\n",
    "            df['Total ' + k] = df[classes].sum().sum()\n",
    "            return df\n",
    "\n",
    "        df = addsum(df, 'Mammalia')\n",
    "        df = addsum(df, 'Sauropsida')\n",
    "        df = addsum(df, 'Actinopteri')\n",
    "        df = df.T.sort_values(kind, ascending=False)\n",
    "\n",
    "        return df\n",
    "    \n",
    "        # stats.whichdup('Convergent', rf, 0.15)\n",
    "    \n",
    "    # All duplication events\n",
    "    def alldups(t, rf):\n",
    "        dfc = stats.whichdup('Convergent', rf, t).reset_index()\n",
    "        dft = stats.whichdup('Tandem', rf, t).reset_index()\n",
    "        dfd = stats.whichdup('Divergent', rf, t).reset_index()\n",
    "\n",
    "        dftc = pd.merge(dft, dfc, on='index')\n",
    "        df = pd.merge(dftc, dfd, on='index')\n",
    "        df = df.set_index('index')\n",
    "        df.index.name = None\n",
    "\n",
    "        return df\n",
    "    \n",
    "        #stats.alldups(0.15, rf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class uni:\n",
    "\n",
    "    def all_scores(kind, threshold):\n",
    "\n",
    "        path = cwd + '/alignments/' + kind + '_alignments/'\n",
    "        file = 'alignments.json'\n",
    "        db = json.load(open(path + file))\n",
    "\n",
    "        scores = []\n",
    "        for l in db:\n",
    "            pos = db[l]['Positions']\n",
    "            for l1 in pos:\n",
    "                s = pos[l1]['Score']\n",
    "                if s > threshold:\n",
    "                    scores.append(s)\n",
    "\n",
    "        return scores\n",
    "    \n",
    "        #uni.all_scores('Convergent', 0)\n",
    "    \n",
    "    def filtdf(kind, threshold):\n",
    "\n",
    "        # scrivo i range di posizioni che mi interessano (in questo caso -1 e +1 sia per begin che per end)\n",
    "        df = pd.read_table(cwd + '/uniprot/' + kind + '_filtered_features.csv', sep=',').drop(columns='Unnamed: 0').reset_index()\n",
    "        df['begin'] = df['begin'].apply(lambda x: None if '~' in str(x) else x)\n",
    "        df['end'] = df['end'].apply(lambda x: None if '~' in str(x) else x)\n",
    "        df['b+1'] = df['begin'].apply(lambda x: str(int(x)+1) if not x == None else x)\n",
    "        df['b-1'] = df['begin'].apply(lambda x: str(int(x)-1) if not x == None else x)\n",
    "        df['e+1'] = df['end'].apply(lambda x: str(int(x)+1) if not x == None else x)\n",
    "        df['e-1'] = df['end'].apply(lambda x: str(int(x)-1) if not x == None else x)\n",
    "        df['b+1'], df['b-1'] = df['b+1'].astype(float), df['b-1'].astype(float)\n",
    "        df['e+1'], df['e-1'] = df['e+1'].astype(float), df['e-1'].astype(float)\n",
    "        df['begin'], df['end'] = df['begin'].astype(float), df['end'].astype(float)\n",
    "\n",
    "        uniprot_list = df[['ensembl_accession', 'index', 'begin', 'b+1', 'b-1', 'end', 'e+1', 'e-1']].values.tolist()\n",
    "\n",
    "        positions = json.load(open(cwd + '/alignments/' + kind + '_alignments/alignments.json'))\n",
    "\n",
    "        # confronto le posizioni trovate in uniprot con quelle trovate mediante allineamenti\n",
    "        prova = []\n",
    "\n",
    "        for group in positions:\n",
    "            x = positions[group]\n",
    "            acc_A, acc_B = x['Accessions'][0], x['Accessions'][1]\n",
    "\n",
    "            for pos in x['Positions']:\n",
    "                x2 = x['Positions'][pos]\n",
    "                score = x['Positions'][pos]['Score']\n",
    "\n",
    "                if score > threshold:\n",
    "                    al_A, al_B = x2['alignment_A'], x2['alignment_B']\n",
    "                    pos_A, pos_B = float(x2['Homo_position_A']), float(x2['Homo_position_B'])\n",
    "                    res_A, res_B = x2['Homo_residue_A'], x2['homo_residue_B']\n",
    "\n",
    "                    for line in uniprot_list:\n",
    "                        accession = line[0]\n",
    "                        index = line[1]\n",
    "                        begin, begin_plus_one, begin_minus_one = line[2], line[3], line[4]\n",
    "                        end, end_plus_one, end_minus_one = line[5], line[6], line[7]\n",
    "\n",
    "                        if accession == acc_A:\n",
    "\n",
    "                            if pos_A == begin or pos_A == begin_plus_one or pos_A == begin_minus_one or pos_A == end or pos_A == end_plus_one or pos_A == end_minus_one:\n",
    "                                prova.append([acc_A, index, pos_A, res_A, res_B, al_A, al_B, score])\n",
    "                            if pos_B == begin or pos_B == begin_plus_one or pos_B == begin_minus_one or pos_B == end or pos_B == end_plus_one or pos_B == end_minus_one:\n",
    "                                prova.append([acc_A, index, pos_B, res_A, res_B, al_A, al_B, score])    \n",
    "\n",
    "                        if accession == acc_B:\n",
    "                            if pos_A == begin or pos_A == begin_plus_one or pos_A == begin_minus_one or pos_A == end or pos_A == end_plus_one or pos_A == end_minus_one:\n",
    "                                prova.append([acc_B, index, pos_A, res_A, res_B, al_A, al_B, score])\n",
    "                            if pos_B == begin or pos_B == begin_plus_one or pos_B == begin_minus_one or pos_B == end or pos_B == end_plus_one or pos_B == end_minus_one:\n",
    "                                prova.append([acc_B, index, pos_B, res_A, res_B, al_A, al_B, score]) \n",
    "\n",
    "        positions_found = pd.DataFrame(prova, columns=['ensembl_accession', 'index', 'position_found', 'residue_A', 'residue_B', 'alignment_A', 'alignment_B', 'score'])\n",
    "\n",
    "        final_df = pd.merge(df, positions_found, on=['index', 'ensembl_accession']).drop_duplicates()\n",
    "\n",
    "        # recupero i full protein name di ensembl dal database (proteina in esame e partner)\n",
    "        database = json.load(open('database.json'))\n",
    "        orthos = pd.read_table(cwd + '/orthologues/' + kind + '_orthologues.csv', sep=';')\n",
    "        orthos = list(zip(orthos[orthos['Species'] == 'Homo_sapiens'].values.tolist()[0][3:], orthos[orthos['Species'] == 'Homo_sapiens'].columns.tolist()[3:]))\n",
    "        orthos = pd.DataFrame(orthos, columns = ['ensembl_accession', 'orthogroup'])\n",
    "        final_df = pd.merge(final_df, orthos, on=['ensembl_accession', 'orthogroup'])\n",
    "        orthos['full_name'] = orthos['ensembl_accession'].apply(lambda x: database.get(x)[7] if not database.get(x) == None else x)\n",
    "\n",
    "        final_df['full_name_ensembl_partner'] = final_df['orthogroup'].apply(lambda x: database.get(\n",
    "            orthos[orthos['orthogroup'] == re.split('A|B', x)[0] + {'A': 'B', 'B': 'A'}.get(x[-1])].values.tolist()[0][0])[7] if not database.get(\n",
    "            orthos[orthos['orthogroup'] == re.split('A|B', x)[0] + {'A': 'B', 'B': 'A'}.get(x[-1])].values.tolist()[0][0]) == None else None)\n",
    "\n",
    "        final_df['full_name_ensembl'] = final_df['orthogroup'].apply(lambda x: database.get(\n",
    "            orthos[orthos['orthogroup'] == x].values.tolist()[0][0])[7] if not database.get(\n",
    "            orthos[orthos['orthogroup'] == x].values.tolist()[0][0]) == None else None)\n",
    "\n",
    "        # ordino la tabella e la salvo in locale\n",
    "        final_df = final_df.drop(columns=['index', 'b+1', 'b-1', 'e+1', 'e-1']).drop_duplicates()[['orthogroup', 'accession', 'ensembl_accession', 'gene', 'full_protein_name', 'full_name_ensembl', 'full_name_ensembl_partner', 'type', 'category', 'description', 'begin', 'end', 'position_found', 'score', 'residue_A', 'residue_B', 'alignment_A', 'alignment_B', 'EC_number']]\n",
    "        final_df = final_df[~final_df['category'].str.contains('STRUCTURAL')]\n",
    "        final_df = final_df[~final_df['category'].str.contains('TOPOLOGY')]\n",
    "\n",
    "        return final_df\n",
    "\n",
    "        #uni.filtdf('Convergent', 0)\n",
    "        \n",
    "    def types(kind, threshold):\n",
    "\n",
    "        if threshold:\n",
    "            df = uni.filtdf(kind, threshold)\n",
    "        else:\n",
    "            df = pd.read_table(cwd + '/uniprot/' + kind + '_features.csv', sep=',')\n",
    "        types = df.groupby('type').count()['accession'].reset_index()\n",
    "        types = types.rename(columns={'accession': kind})\n",
    "        u_types = df.drop_duplicates(['type', 'accession'])\n",
    "        u_types = u_types.groupby('type').count()['accession'].reset_index()\n",
    "        u_types = u_types.rename(columns={'accession': 'Unique_' + kind.lower()})\n",
    "        df = pd.merge(types, u_types, on='type')\n",
    "\n",
    "        return df\n",
    "    \n",
    "        #uni.types('Convergent', 0)\n",
    "        \n",
    "    def allfeat(threshold):\n",
    "\n",
    "        p = pd.merge(uni.types('Tandem', threshold), uni.types('Convergent', threshold), on='type', how='outer')\n",
    "        df = pd.merge(p, uni.types('Divergent', threshold), on='type', how='outer').set_index('type')\n",
    "        df.index.name = None\n",
    "\n",
    "        return df\n",
    "\n",
    "        #uni.allfeat(None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
